packages/data-provider/src/services/submissionService.ts

```diff
@@ -174,6 +150,112 @@ const service = (dependencies: BaseDependencies) => {
+		const dataMergedByEntityName = mergeActiveSubmissionAndSubmittedData(submittedData, {
+			data: newSubmissionData,
+			id: originalSubmission.id,
+		});
+
+		const currentDictionary = await getActiveDictionaryByCategory(originalSubmission.dictionaryCategoryId);
+		if (!currentDictionary) {
+			throw new BadRequest(`Dictionary in category '${originalSubmission.dictionaryCategoryId}' not found`);
+		}
+
+		// Prepare data to validate. Extract schema data from merged data
+		const crossSchemasDataToValidate = extractSchemaDataFromMergedDataRecords(dataMergedByEntityName);
+
+		// Run validation using Lectern Client
+		const resultValidation = validateSchemas(currentDictionary, crossSchemasDataToValidate);
+
+		// Collect errors of the Active Submission
+		const submissionSchemaErrors = groupSchemaErrorsByEntity({
+			resultValidation,
+			dataValidated: dataMergedByEntityName,
+		});
+
+		if (_.isEmpty(submissionSchemaErrors)) {
+			logger.info(LOG_MODULE, `No error found on data submission`);
+		}
+
+		// Update Active Submission
+		return await createOrUpdateActiveSubmission({
+			idActiveSubmission: originalSubmission.id,
+			entityMap: newSubmissionData,
+			categoryId: originalSubmission.dictionaryCategoryId.toString(),
+			schemaErrors: submissionSchemaErrors,
+			dictionaryId: currentDictionary.id,
+			userName: userName,
+			organization: originalSubmission.organization,
+		});
+	};
+
+	/**
+	 * Void function to process and validate uploaded files on an Active Submission.
+	 * Performs the schema data validation combined with all Submitted Data.
+	 * @param {Record<string, Express.Multer.File>} files Uploaded files to be processed
+	 * @param {Object} params
+	 * @param {number} params.categoryId Category Identifier
+	 * @param {string} params.organization Organization name
+	 * @param {string} params.userName User who performs the action
+	 * @returns {void}
+	 */
+	const validateFilesAsync = async (files: Record<string, Express.Multer.File>, params: ValidateFilesParams) => {
+		const { getActiveSubmission } = submissionRepository(dependencies);
+		const { submissionEntitiesFromFiles } = submissionUtils(dependencies);
+
+		const { categoryId, organization, userName } = params;
+
+		// Parse file data
+		const filesDataProcessed = await submissionEntitiesFromFiles(files, userName);
+
+		// Get Active Submission from database
+		const activeSubmission = await getActiveSubmission({ categoryId, userName, organization });
+		if (!activeSubmission) {
+			throw new BadRequest(`Submission '${activeSubmission}' not found`);
+		}
+
+		// Merge Active Submission data with incoming TSV file data processed
+		const updatedActiveSubmissionData: Record<string, SubmissionData> = {
+			...activeSubmission.data,
+			...filesDataProcessed,
+		};
+
+		// Perform Schema Data validation Async.
+		performDataValidation({
+			originalSubmission: activeSubmission,
+			newSubmissionData: updatedActiveSubmissionData,
+			userName,
+		});
```

Comment:
Wonder if this is a good worker threads candidate. If the validation is slow and thread blocking this could be useful to move off the main thread. Consider opening a new issue to make that change.

Reply:
ticket #51 created to track this change